### PR TITLE
Move function typed properties from the SourceFile to a dedicated functions

### PR DESCRIPTION
**Pull request by vladima**
Function typed properties on SourceFile are made with closures, for all source files it will keep the parser instance alive. This PR makes SourceFile pure data object -> no closures are required

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -356,6 +356,368 @@ module ts {
...
+    // Produces a new SourceFile for the 'newText' provided. The 'textChangeRange' parameter 
+    // indicates what changed between the 'text' that this SourceFile has and the 'newText'.
+    // The SourceFile will be created with the compiler attempting to reuse as many nodes from 
+    // this file as possible.
+    //
+    // Note: this function mutates nodes from this SourceFile. That means any existing nodes
+    // from this SourceFile that are being held onto may change as a result (including 
+    // becoming detached from any SourceFile).  It is recommended that this SourceFile not
+    // be used once 'update' is called on it.
+    export function update(sourceFile: SourceFile, newText: string, textChangeRange: TextChangeRange): SourceFile {
```
Comment (ahejlsberg) [inline src/compiler/parser.ts]: Maybe `updateSourceFile`?

**Comment by ahejlsberg** [comment]
Looks good, but hard to see exactly what changed in the large blocks that moved.

File: src/compiler/parser.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/parser.ts]: will change

**Comment by vladima** [comment]
Large blocks are mostly auxiliary function that were used by `update`. Previously these functions were nested in `createSourceFile` and now they were pushed in the outer scope

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -391,7 +391,7 @@ module FourSlash {
             this.currentCaretPosition = pos;
 
             var lineStarts = ts.computeLineStarts(this.getCurrentFileContent());
-            var lineCharPos = ts.getLineAndCharacterOfPosition(lineStarts, pos);
+            var lineCharPos = ts.computeLineAndCharacterOfPosition(lineStarts, pos);
```
Comment (JsonFreeman) [inline src/harness/fourslash.ts]: What's the difference between this and `getLineAndCharacterOfPosition`?

Comment (vladima) [inline src/harness/fourslash.ts]: `computeLineAndCharacterOfPosition` accepts lineMap as `number[]` and use it to compute line and character. `getLineAndCharacterOfPosition` accepts `SourceFile`, picks lineMap from the file (and initializes it if it was not yet created) and then invokes `computeLineAndCharacterOfPosition`

File: src/compiler/scanner.ts [hunk 3]
```diff
@@ -278,12 +278,20 @@ module ts {
...
         Debug.assert(line > 0);
         return lineStarts[line - 1] + character - 1;
     }
 
-    export function getLineAndCharacterOfPosition(lineStarts: number[], position: number) {
+    export function getLineStarts(sourceFile: SourceFile): number[] {
+        return sourceFile.lineMap || (sourceFile.lineMap = computeLineStarts(sourceFile.text));
+    }
+
+    export function computeLineAndCharacterOfPosition(lineStarts: number[], position: number) {
```
Comment (DanielRosenwasser) [inline src/compiler/scanner.ts]: For anything new functions introduced, don't export them from `scanner.ts`, expose them from `utilities.ts`.

@mhegazy, this relates to the API test we need to reintroduce; Monday?

I'd also annotate the return types, but that's just me.

File: src/compiler/types.ts [hunk 4]
```diff
@@ -907,12 +892,15 @@ module ts {
         // missing tokens, or tokens it didn't know how to deal with).
         parseDiagnostics: Diagnostic[];
 
-        // Returns all syntactic diagnostics (i.e. the reference, parser and grammar diagnostics).
-        getSyntacticDiagnostics(): Diagnostic[];
+        //getSyntacticDiagnostics(): Diagnostic[];
```
Comment (mhegazy) [inline src/compiler/types.ts]: remove this commented out line.

**Comment by sheetalkamat** [comment]
:+1:

**Pull request merged**